crates/volta-core/src/tool/package/fetch.rs

```diff
@@ -0,0 +1,143 @@
+//! Provides fetcher for 3rd-party packages
+
+use std::fs::{rename, write, File};
+use std::io::{Read, Seek, SeekFrom};
+use std::path::{Path, PathBuf};
+
+use super::super::download_tool_error;
+use crate::error::ErrorDetails;
+use crate::fs::{
+    create_staging_dir, ensure_containing_dir_exists, ensure_dir_does_not_exist, read_dir_eager,
+    read_file,
+};
+use crate::path;
+use crate::style::{progress_bar, tool_version};
+use crate::tool::{self, PackageDetails};
+use crate::version::VersionSpec;
+use archive::{Archive, Tarball};
+use log::debug;
+use semver::Version;
+use sha1::{Digest, Sha1};
+use volta_fail::{Fallible, ResultExt};
+
+pub fn fetch(name: &str, details: &PackageDetails) -> Fallible<()> {
+    let version_string = details.version.to_string();
+    let cache_file = path::package_distro_file(&name, &version_string)?;
+    let shasum_file = path::package_distro_shasum(&name, &version_string)?;
+
+    if let Some(archive) = load_cached_distro(&cache_file, &shasum_file) {
```

Comment:
Possible alternative here which I think might be clearer:

```rust
let (archive, cached) = match load_cached_distro(&cache_file, &shasum_file) {
    Some(archive) => (archive, true),
    None => fetch_remote_distro(
        tool::Spec::Package(name.into(), VersionSpec::exact(&details.version)),
        &details.tarball_url,
        &cache_file,
    )
    .map(|archive| (archive, false))?,
};

if cached {
    debug!(
        "Loading {} from cached archive at '{}'",
        tool_version(&name, &version_string),
        cache_file.display(),
    );
}

unpack_archive(archive, name, &details.version)?;

if cached {
    Ok(())
} else {
    write(&shasum_file, details.shasum.as_bytes()).with_context(|_| {
        ErrorDetails::WritePackageShasumError {
            package: name.into(),
            version: version_string,
            file: shasum_file,
        }
    })
}
```

As for *why* I think that's clearer: it seems to me that it more cleanly separates the concerns of the function into their own distinct blobs; it's clearer here how you'd factor it out into its own function if you wanted.

Reply:
I like that approach. I actually implemented it slightly differently, including the `debug!` message as part of the `Some(archive)` match arm, but that cleans it up nicely and makes the flow clearer, thanks!

Reply:
That makes sense and is probably cleaner!